src/main.rs

```diff
@@ -53,30 +157,69 @@ fn main() {
+        // and <filter_name>.rs for the actual filter implementation
+        
+        // Making new library folder
+        let mut lib_src_folder = filter_name.to_string();
+        lib_src_folder.push_str("/src/");
+        fs::create_dir_all(&lib_src_folder).unwrap();
 
-    let handlebars = Handlebars::new();
+        // Cargo.toml
+        let mut cargo_file_name = filter_name.clone();
+        cargo_file_name.push_str("/");
+        cargo_file_name.push_str("/Cargo.toml");
+        let mut cargo_handlebars_name = TEMPLATE_FOLDER.to_string();
+        cargo_handlebars_name.push_str("Cargo.toml.handlebars");
+        generate_code_from_filter_name_with_handlebars(filter_name.clone(), &cargo_handlebars_name, cargo_file_name);
 
-    let output = handlebars
-        .render_template(&template_str, &code_gen)
-        .expect("handlebar render failed");
+        // Filter types
+        let mut filter_types_file = filter_name.clone();
+        filter_types_file.push_str("/");
+        filter_types_file.push_str(&filter_name);
+        filter_types_file.push_str("_types.rs");
+        let mut filter_types_handlebars = TEMPLATE_FOLDER.to_string();
+        filter_types_handlebars.push_str("src/filter_types.rs.handlebars");
+        generate_code_from_codegen_with_handlebars(&code_gen, &filter_types_handlebars, filter_types_file);
 
-    let mut file = File::create("./cpp_filter/filter.cc").expect("file create failed.");
-    file.write_all(output.as_bytes()).expect("write failed");
+        // The lib file
+        let mut lib_file_name = filter_name.clone();
+        lib_file_name.push_str("/src/lib.rs");
+        let mut lib_file_handlebars = TEMPLATE_FOLDER.to_string();
+        lib_file_handlebars.push_str("src/lib.rs.handlebars");
+        generate_code_from_filter_name_with_handlebars(filter_name.clone(), &lib_file_handlebars, lib_file_name);
+        
+        // The filter itself
+        let mut filter_file_name = filter_name.clone();
+        filter_file_name.push_str("/src/");
+        filter_file_name.push_str(&filter_name);
+        filter_file_name.push_str(".rs");
+        let mut filter_name_handlebars = TEMPLATE_FOLDER.to_string();
+        filter_name_handlebars.push_str("src/filter.rs.handlebars");
+        generate_code_from_codegen_with_handlebars(&code_gen, &filter_name_handlebars, filter_file_name);
+
+        // The graph_utils file
+        let mut graph_utils_new_file_name = filter_name.clone();
+        graph_utils_new_file_name.push_str("/src/graph_utils.rs");
+        let mut graph_utils_old_file_name = TEMPLATE_FOLDER.to_string();
+        graph_utils_old_file_name.push_str("src/graph_utils.rs");
+        fs::copy(&graph_utils_old_file_name, &graph_utils_new_file_name).unwrap();
+
+
+    }
+    else {
+        let mut filter_name_cc = filter_name.to_string().clone();
+        filter_name_cc.push_str(".cc");
```

Comment:
The Path object in Rust provides a function called [set_extension](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.set_extension). I would use that.

Reply:
That's a useful library.  I changed all path handling to use it.